third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1186 @@
+//
+//   - `rewrite_local_var` replaces symbol uses that are shadowed with the shadow symbol.
+//   - `rewrite_temporary` replaces references to function parameters with parameter symbols, also shadowing as needed.
+//
+//   - `rewrite_invoke` handles calls to lambda parameters within the inlined function
+//      - Note that lambda bodies are not rewritten during inlining, but are kept intact
+//
+//   - `rewrite_pattern` replaces syms in a pattern with shadow symbols as necessary
+//
+// - struct `InlinedRewriter` also has various methods to support rewriting a call
+//   - `inline_call` is the entry point for rewriting a call to an inline function.
+//
+//   - `create_shadow_symbols` creates a shadow `Symbol` for each lambda free variable to use if
+//      a local variable conflicts with that variable.  It uses `SymbolPool::shadow` to create
+//      shadow symbols which print out the same but don't conflict.
+//
+//   - `get_shadow_symbol` checks whether a `Symbol` conflicts with lambda free variables, and if so,
+//     returns the shadow symbol to use instead..
+//
+//   - `parameter_list_to_pattern` is a helper to convert a list of `Parameter` into a `Pattern`,
+//      suitable for use in the body.
+//
+//   - `construct_inlined_call_expression` is a helper to build the expression corresponding to
+//      { let params=actuals; body } used for both lambda inlining and inline function inlining.
+//
+//   - rewrite_pattern_vector is a helper for `rewrite_pattern`
+//
+
+// TODO:
+// - add a InlinedCall() node that represents an inlined call, so that an inlined Return() has a place to go. [later]
+//   - OR: if an inlined fn has a return, then error out.
+// - add a simplifier that simplifies certain code constructs, e.g.
+//   - InlinedCall() with no nested Return() can be flattened.
+//   - others?
+// - do we need to insert FreezeRef in actual args if formal param is &T but arg is &mod T ?
+// - do we need to do anything about abilities?
+// - if lambda parameters also may be referred to by temporaries, then `rewrite_invoke` might need to call yet
+//   another `ExpRewriterFunctions` implementation to do that.
+
+use codespan_reporting::diagnostic::Severity;
+use itertools::chain;
+use move_model::{
+    ast::{Exp, ExpData, Operation, Pattern, TempIndex},
+    exp_rewriter::ExpRewriterFunctions,
+    model::{FunId, GlobalEnv, Loc, NodeId, Parameter, QualifiedId, TypeParameter},
+    symbol::Symbol,
+    ty::{ReferenceKind, Type},
+};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fmt::Debug,
+    iter::{zip, IntoIterator, Iterator},
+    ops::Deref,
+    vec::Vec,
+};
+
+// ======================================================================================
+// Entry
+
+// Run inlining on current program's AST.  For each function which is target of the compilation,
+// visit that function body and inline any calls to functions marked as "inline".
+pub fn run_inlining(env: &mut GlobalEnv, debug: bool) {
```

Comment:
OK, probably should move to GlobalEnv eventually, but not at the moment.

Reply:
The canonical way is to put this into `Options`. The instance of options on which the compiler operates is [stored the in env after type checlk](https://github.com/aptos-labs/aptos-core/blob/530df40a8a2cd357f4467d50400399bacace0a42/third_party/move/move-compiler-v2/src/lib.rs#L101C1-L102), and can be retrieved with `env.get_extension::<Options>()`. 

Notice that the `Options` type can be nicely declaratively configured for command flags. If you want an environment variable, you can specify this via a clap attribute that an option is gotten from there. However, in general, `Options` is not tied to command flags but the central configuration point.

Can we do this already now, or is something blocking?

Reply:
Well, that's an interesting approach.  It might be safer to have an actual field in the `GlobalEnv` struct so it's a little more discoverable and we could be sure it's set on every path.  But I can pull debug out of there now that I know it is there.  I'll try that tomorrow.

Reply:
Done.